src/controller/CommissioningWindowOpener.h

```diff
@@ -104,8 +106,8 @@ class CommissioningWindowOpener
      */
     CHIP_ERROR OpenCommissioningWindow(NodeId deviceId, System::Clock::Seconds16 timeout, uint32_t iteration,
                                        uint16_t discriminator, Optional<uint32_t> setupPIN, Optional<ByteSpan> salt,
-                                       Callback::Callback<OnOpenCommissioningWindow> * callback, SetupPayload & payload,
-                                       bool readVIDPIDAttributes = false);
+                                       Optional<ByteSpan> verifier, Callback::Callback<OnOpenCommissioningWindow> * callback,
```

Comment:
This makes no sense.  There is never a situation in which you would both pass in the PIN _and_ pass in the verifier, so why do we have a method that lets you pass both?

It sounds like we need to have two separate methods, and the impl gets the verifier (provided or generated) and then calls a common impl method.

Reply:
With the params refactor, here's the new shape of the two functions I'm thinking of, let me know your feedback

```cpp

    struct CommissioningWindowParams
    {
        // The device Id of the node.
        NodeId deviceId;
        // The duration for which the commissioning window should remain open.
        System::Clock::Seconds16 timeout;
        // The PAKE iteration count associated with the PAKE Passcode ID and ephemeral PAKE passcode verifier to be used for this
        // commissioning.
        uint32_t iteration;
        // The long discriminator for the DNS-SD advertisement.
        uint16_t discriminator;
        // The salt to use, or NullOptional to use a randomly-generated one. If provided, must be at least
        // kSpake2p_Min_PBKDF_Salt_Length bytes and at most kSpake2p_Max_PBKDF_Salt_Length bytes in length.
        Optional<ByteSpan> salt = NullOptional;
        // Should the API internally read VID and PID from the device while opening the commissioning window.  If this argument is
        // `true`, the API will read VID and PID from the device and include them in the setup payload passed to the callback.
        bool readVIDPIDAttributes = false;
        Callback::Callback<OnOpenCommissioningWindow> *
            callback; // The function to be called on success or failure of opening the commissioning window.
    };

    /**
     * @brief
     *   Try to look up the device attached to our controller with the given
     *   node id and ask it to re-enter commissioning mode with a PASE verifier
     *   derived from the given information and the given discriminator. The
     *   device will exit commissioning mode after a successful commissioning,
     *   or after the given `timeout` time.
     *
     * @param[in] params        The common parameters required to open the commissioning window.
     * @param[in] setupPIN      The setup PIN to use, or NullOptional to use a randomly-generated one.
     * @param[out] payload      The setup payload, not including the VID/PID bits,
     *                          even if those were asked for, that is generated
     *                          based on the passed-in information.  The payload
     *                          provided to the callback function, unlike this
     *                          out parameter, will include the VID/PID bits if
     *                          readVIDPIDAttributes is true.
     */
    CHIP_ERROR OpenCommissioningWindow(const CommissioningWindowParams & params, Optional<uint32_t> setupPIN,
                                       SetupPayload & payload);

    /**
     * @brief
     *   Try to look up the device attached to our controller with the given
     *   node id and ask it to re-enter commissioning mode with a PASE verifier
     *   derived from the given information and the given discriminator. The
     *   device will exit commissioning mode after a successful commissioning,
     *   or after the given `timeout` time.
     *
     * @param[in] params        The common parameters required to open the commissioning window.
     * @param[in] verifier      The PAKE passcode verifier to use, or NullOptional to generate verifier based on randomly generated
     *                          PIN and other parameters.
     * @param[out] payload      The setup payload, not including the VID/PID bits,
     *                          even if those were asked for, that is generated
     *                          based on the passed-in information.  The payload
     *                          provided to the callback function, unlike this
     *                          out parameter, will include the VID/PID bits if
     *                          readVIDPIDAttributes is true.
     */
    CHIP_ERROR OpenCommissioningWindow(const CommissioningWindowParams & params, Optional<ByteSpan> verifier,
                                       SetupPayload & payload);
```

Reply:
I think there are a few issues.  First, optional for salt (which means "use a random one" if not provided) does not make sense with a verifier, because the salt needs to match the verifier.  

I think there are two ways of presenting the "things verifier will be derived from":

1. Iteration count, optional passcode (mis-named PIN in this API), optional salt.
2. Iteration count, salt, verifier that corresponds to the provided iteration count and salt and some not-provided passcode.

In terms of modeling this in the API, seems like the cleanest thing are two separate object classes to encapsulate those two separate pieces of information, right?  Whether then we do some sort of static dispatch via overloads or have them inherit from a class that has a pure virtual method that outputs the iteration count, salt, and verifier is something to think about.

Second issue: there is no way to create a SetupPayload from a verifier: the whole point of a SetupPayload is to know the passcode so you can do PASE.  So the API version that takes a verifier can't have a SetupPayload outparam.

Third issue: The version of the API with a verifier can't pass a SetupPayload to the callback function, for exactly the same reasons.

Reply:
Thanks for detailing these. Indeed I came across the second issue during the refactor and that led to digging in for the third one which seems to be even more problematic.

For the first one, I'm leaning towards just taking the salt out of common params class, and making it part of first API (with passcode) as optional, and second one (with verifier) as required.

Second: the setup payload doesn't need to be an outparam for the second API, and the current consumer of it (fabric-admin) just ignores that anyway so we can safely remove this.

Third: This is where I'm torn between either introducing two different callback functions or just making the setup payload optional in the existing one. Essentially either
```cpp
typedef void (*OnOpenCommissioningWindowWithPasscode)(void * context, NodeId deviceId, CHIP_ERROR status, SetupPayload payload);
typedef void (*OnOpenCommissioningWindowWithVerifier)(void * context, NodeId deviceId, CHIP_ERROR status);
typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
```
(we could also keep the default `OnOpenCommissioningWindowWithPasscode` as `OnOpenCommissioningWindow` to keep the diff small)

or
```cpp
typedef void (*OnOpenCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status, Optional<SetupPayload payload>);
typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
```

Callback function can then also be taken out of common params depending on the solution opted for (3)

Reply:
At that point, it seems like we might just want to have two different params classes (which perhaps have a common superclass for the shared bits if desired).  Having different callback function types for the two cases makes the most sense to me.

Keeping the existing API working on top of the new one seems pretty easy and would reduce the amount of code churn SDK clients have to have.

Reply:
Addressed these changes in https://github.com/project-chip/connectedhomeip/pull/34209/commits/5d4ce206b24d3a3a4201e6d1cdc816ca6e7f8b15, LMK if there's anything else missing or if it should be done differently.